schema/query.go

```diff
@@ -107,6 +121,56 @@ func validateQuery(q map[string]interface{}, validator Validator, parentKey stri
 	queries := Query{}
 	for key, exp := range q {
 		switch key {
+		case "$regex":
+			if parentKey == "" {
+				return nil, errors.New("$regex can't be at first level")
+			}
+
+			switch exp.(type) {
+			case string:
+				regex := exp.(string)
+				v, err := regexp.Compile(regex)
+				if err != nil {
+					return nil, fmt.Errorf("$regex: invalid regex: %v", err)
+				}
+
+				value := RegexValue{
+					RegValue:   v,
+					RegOptions: "",
+				}
+
+				queries = append(queries, Regex{Field: parentKey, Value: value})
+			case []interface{}:
+				regex := exp.([]interface{})
+
+				var (
+					v   *regexp.Regexp
+					err error
+				)
+				// Control regex options behavour.
+				// First check if we find any options of the regexp package
+				// Supported options (flags):
+				// i              case-insensitive (default false)
+				// m              multi-line mode: ^ and $ match begin/end line in addition to begin/end text (default false)
+				// s              let . match \n (default false)
+				// U              ungreedy: swap meaning of x* and x*?, x+ and x+?, etc (default false)
+				if strings.ContainsAny(regex[1].(string), "imsU") {
```

Comment:
As it stands, this line may panic with index out of range. Also the .(string) typecast may result in a panic when ok is not checked.

I suggest you extract regex[0] and optionally regex[1] into two local variables while handling errors appropriately.

Other than that I assume ContainsAny won't check if regex[1] contains any other strings than imsU, so if I where to supply "imsUxab", would the regex.Comile then fail? Would perhaps be good to have a unittest for this:-)